File: sample_app/lib/services/mcp_client.dart
import 'dart:async';
import 'dart:convert';

import 'package:web_socket_channel/web_socket_channel.dart';

/// Минимальный интерфейс клиента MCP для мокирования в тестах.
abstract class McpApi {
  Future<dynamic> toolsCall(String name, Map<String, dynamic> args, {Duration? timeout});
}

class McpClient implements McpApi {
  WebSocketChannel? _channel;
  StreamSubscription? _sub;
  int _idCounter = 1;
  final Map<int, Completer<dynamic>> _pending = {};
  bool _initialized = false;

  bool get isConnected => _channel != null;
  bool get isInitialized => _initialized;

  Future<void> connect(String url) async {
    await close();
    _channel = WebSocketChannel.connect(Uri.parse(url));
    _sub = _channel!.stream.listen(_onMessage, onError: (e) {
      // fail all pending
      for (final c in _pending.values) {
        if (!c.isCompleted) c.completeError(e);
      }
      _pending.clear();
    }, onDone: () {
      for (final c in _pending.values) {
        if (!c.isCompleted) c.completeError(StateError('WebSocket closed'));
      }
      _pending.clear();
      _initialized = false;
      _channel = null;
    });
  }

  Future<void> close() async {
    await _sub?.cancel();
    _sub = null;
    await _channel?.sink.close();
    _channel = null;
    _initialized = false;
  }

  void _onMessage(dynamic raw) {
    try {
      final msg = jsonDecode(raw as String) as Map<String, dynamic>;
      final id = msg['id'];
      if (id is int && _pending.containsKey(id)) {
        final c = _pending.remove(id)!;
        if (msg.containsKey('error')) {
          c.completeError(msg['error']);
        } else {
          c.complete(msg['result']);
        }
      }
    } catch (_) {
      // ignore malformed
    }
  }

  Future<dynamic> _send(
    String method, {
    Map<String, dynamic>? params,
    Duration? timeout,
  }) {
    if (_channel == null) {
      throw StateError('MCP not connected');
    }
    final id = _idCounter++;
    final c = Completer<dynamic>();
    _pending[id] = c;
    final payload = {
      'jsonrpc': '2.0',
      'id': id,
      'method': method,
      if (params != null) 'params': params,
    };
    _channel!.sink.add(jsonEncode(payload));
    return c.future.timeout(timeout ?? const Duration(seconds: 15));
  }

  Future<Map<String, dynamic>> initialize({Duration? timeout}) async {
    final result = await _send('initialize', timeout: timeout) as Map<String, dynamic>;
    _initialized = true;
    return result;
  }

  Future<Map<String, dynamic>> toolsList({Duration? timeout}) async {
    final result = await _send('tools/list', timeout: timeout) as Map<String, dynamic>;
    return result;
  }

  @override
  Future<dynamic> toolsCall(String name, Map<String, dynamic> args, {Duration? timeout}) async {
    final result = await _send(
      'tools/call',
      params: {
        'name': name,
        'arguments': args,
      },
      timeout: timeout,
    );
    return result;
  }
}

File: issue-monitor/src/main/kotlin/ru/marslab/issuemonitor/mcp/McpClient.kt
package ru.marslab.issuemonitor.mcp

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import okhttp3.*
import okio.ByteString
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

class McpClient(private val url: String) : WebSocketListener() {
    private val client = OkHttpClient.Builder()
        .pingInterval(20, TimeUnit.SECONDS)
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()

    private var webSocket: WebSocket? = null

    private val json = Json { ignoreUnknownKeys = true }
    private val pending = ConcurrentHashMap<Long, CompletableDeferred<JsonObject>>()
    private val idMutex = Mutex()
    private var lastId = 0L

    suspend fun connect() {
        if (webSocket != null) return
        val request = Request.Builder().url(url).build()
        webSocket = client.newWebSocket(request, this)
    }

    suspend fun ensureConnected() {
        if (webSocket == null) connect()
    }

    fun close() {
        webSocket?.close(1000, "bye")
        webSocket = null
    }

    private suspend fun nextId(): Long = idMutex.withLock { ++lastId }

    private fun send(payload: String) {
        val ws = webSocket ?: throw IllegalStateException("WebSocket is not connected")
        ws.send(payload)
    }

    suspend fun initialize(): JsonObject {
        val id = nextId()
        val obj = buildJsonObject {
            put("jsonrpc", "2.0")
            put("id", id)
            put("method", "initialize")
        }
        val deferred = CompletableDeferred<JsonObject>()
        pending[id] = deferred
        send(json.encodeToString(JsonObject.serializer(), obj))
        return deferred.await()
    }

    suspend fun toolsCall(name: String, args: JsonObject): JsonElement {
        val id = nextId()
        val params = buildJsonObject {
            put("name", name)
            put("arguments", args)
        }
        val obj = buildJsonObject {
            put("jsonrpc", "2.0")
            put("id", id)
            put("method", "tools/call")
            put("params", params)
        }
        val deferred = CompletableDeferred<JsonObject>()
        pending[id] = deferred
        send(json.encodeToString(JsonObject.serializer(), obj))
        val resultEnvelope = deferred.await()
        // Expected: { name, result }
        return resultEnvelope["result"] ?: JsonNull
    }

    override fun onOpen(webSocket: WebSocket, response: Response) {
        // no-op
    }

    override fun onMessage(webSocket: WebSocket, text: String) {
        val el = runCatching { json.parseToJsonElement(text) }.getOrNull() ?: return
        val obj = el.jsonObject
        val idEl = obj["id"] ?: return
        val id = when (idEl) {
            is JsonPrimitive -> if (idEl.isString) idEl.content.toLongOrNull() else idEl.longOrNull
            else -> null
        } ?: return

        val result = obj["result"]?.jsonObject
        val error = obj["error"]

        val deferred = pending.remove(id)
        if (deferred != null) {
            if (error != null) {
                deferred.completeExceptionally(RuntimeException("MCP error: $error"))
            } else if (result != null) {
                deferred.complete(result)
            } else {
                deferred.completeExceptionally(RuntimeException("Invalid MCP response: $text"))
            }
        }
    }

    override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
        // ignore binary
    }

    override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
        // Bubble up failures to all waiters
        val ex = RuntimeException("WebSocket failure", t)
        pending.values.forEach { it.completeExceptionally(ex) }
        pending.clear()
        this.webSocket = null
    }

    override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
        // Complete all pending with error
        val ex = RuntimeException("WebSocket closed: $code $reason")
        pending.values.forEach { it.completeExceptionally(ex) }
        pending.clear()
        this.webSocket = null
    }
}
